ledger/participant-integration-api/src/main/scala/platform/store/utils/ConcurrencyLimiter.scala

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.store.utils
+
+import java.util.concurrent.Executors
+import scala.collection.mutable
+import scala.concurrent.duration.Duration
+import scala.concurrent.{Await, ExecutionContext, Future, Promise}
+
+trait ConcurrencyLimiter[T] {
+  def execute(task: () => Future[T]): Future[T]
+}
+
+class NoConcurrencyLimiter[T] extends ConcurrencyLimiter[T] {
+  override def execute(task: () => Future[T]): Future[T] = task()
+}
+
+class QueueBasedConcurrencyLimiter[T](
+    parallelism: Int,
+    executionContext: ExecutionContext,
+) extends ConcurrencyLimiter[T] {
+  assert(parallelism > 0)
+
+  case class WaitingTask(
```

Comment:
You do not really need so much data to track. We could as good as do line 44 and line 52 as we compose the future generating function at line 34. You do not need to have a future generating function neither, just something to create the future and being done with it. So the queue-s signature can be () => Unit.

Reply:
Refactored the code to make the class non-parametric.